src/meta/src/manager/scale.rs

```diff
@@ -0,0 +1,348 @@
+use tokio::sync::{mpsc, Mutex};
+use tokio::task::JoinHandle;
+
+use crate::model::{MetadataModel, ScaleTaskId};
+use crate::storage::MetaStore;
+
+pub type ScaleManagerRef<S> = Arc<ScaleManager<S>>;
+type ScaleTaskCacheRef = Arc<Mutex<HashMap<ScaleTaskId, ScaleTask>>>;
+
+/// [`ScaleManager`] is used to receive scale tasks.
+#[allow(dead_code)]
+pub struct ScaleManager<S: MetaStore> {
+    meta_store: Arc<S>,
+
+    next_task_id: AtomicU32,
+    task_cache: ScaleTaskCacheRef,
+    task_queue_tx: UnboundedSender<ScaleTaskId>,
+}
+
+impl<S> ScaleManager<S>
+where
+    S: MetaStore,
+{
+    pub async fn new(meta_store: Arc<S>) -> Result<(Self, JoinHandle<()>, Sender<()>)> {
+        let tasks = ScaleTask::list(&*meta_store).await?;
+        let (task_queue_tx, mut task_queue_rx) = mpsc::unbounded_channel();
+
+        let next_task_id = tasks.iter().map(|task| task.task_id).max().unwrap_or(0) + 1;
+        tasks
+            .iter()
+            .filter(|task| task.task_status == TaskStatus::Pending as i32)
+            .map(|task| task.task_id)
+            .sorted()
+            .for_each(|task_id| {
+                task_queue_tx.send(task_id).unwrap();
+            });
+        let task_cache = Arc::new(Mutex::new(
+            tasks.into_iter().map(|task| (task.task_id, task)).collect(),
+        ));
+
+        let scale_manager = ScaleManager {
+            meta_store: meta_store.clone(),
+            next_task_id: AtomicU32::new(next_task_id),
+            task_cache: task_cache.clone(),
+            task_queue_tx,
+        };
+
+        let (shutdown_tx, mut shutdown_rx) = tokio::sync::oneshot::channel();
+        let join_handle = tokio::spawn(async move {
+            loop {
+                let task_id = tokio::select! {
+                    task_id = task_queue_rx.recv() => {
+                        match task_id {
+                            None => return,
+                            Some(task_id) => task_id
+                        }
+                    }
+                    _ = &mut shutdown_rx => {
+                        tracing::info!("Scale Manager is stopped");
+                        return;
+                    }
+                };
+                if let Err(err) =
+                    Self::solve_task(meta_store.clone(), task_cache.clone(), task_id).await
+                {
+                    error!("Failed to solve scale task: {}", err);
+                }
+            }
+        });
+
+        Ok((scale_manager, join_handle, shutdown_tx))
+    }
+
+    async fn solve_task(
+        meta_store: Arc<S>,
+        task_cache: ScaleTaskCacheRef,
+        task_id: ScaleTaskId,
+    ) -> Result<()> {
+        let mut task_cache_guard = task_cache.lock().await;
+        let mut task = task_cache_guard.get_mut(&task_id).unwrap();
+
+        if TaskStatus::Cancelled == task.get_task_status()? {
+            return Ok(());
+        }
+
+        task.task_status = TaskStatus::Building as i32;
```

Comment:
Should set task status to `Building` when all dependencies are ready and scale task is really under doing, and task in `Building` status should not be abortable.